appserver/handle_topology.go

```diff
@@ -0,0 +1,531 @@
+		if element.Kind == r.Kind {
+			nMap.nodes[name].nd.Resources[index] = r
+			return
+		}
+	}
+	node := nMap.nodes[name]
+	resources := node.nd.Resources
+	resources = append(resources, r)
+	node.nd.Resources = resources
+	nMap.nodes[name] = node
+}
+
+// Converts the HTTP connection to a websocket in order to stream data.
+func convertHTTPToWebSocket(w http.ResponseWriter, r *http.Request) *websocket.Conn {
+	upgrader.CheckOrigin = func(r *http.Request) bool { return true }
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	return ws
+}
+
+// Create a watcher for Deployments and DeploymentConfigs.
+func createNodeWatcher(namespace string, k *kubeclient.KubeClient) *watcher.Watch {
+	listOptions := metav1.ListOptions{}
+	onGetWatchError := func(err error) {
+		fmt.Errorf("Error is %+v", err)
+	}
+	newWatch := watcher.NewWatch(namespace,
+		k,
+		k.GetDeploymentConfigWatcher(namespace, listOptions, onGetWatchError),
+		k.GetDeploymentWatcher(namespace, listOptions, onGetWatchError),
+	)
+	newWatch.SetFilters([]watch.EventType{watch.Added, watch.Modified, watch.Deleted})
+	newWatch.StartWatcher()
+
+	return newWatch
+}
+
+// Create a watcher for all resources.
+func createResourceWatcher(namespace string, opts metav1.ListOptions, k *kubeclient.KubeClient) *watcher.Watch {
+	onGetWatchError := func(err error) {
+		fmt.Errorf("Error is %+v", err)
+	}
+	newWatch := watcher.NewWatch(namespace,
+		k,
+		k.GetDeploymentConfigWatcher(namespace, opts, onGetWatchError),
+		k.GetDeploymentWatcher(namespace, opts, onGetWatchError),
+		k.GetReplicationControllerWatcher(namespace, opts, onGetWatchError),
+		k.GetReplicaSetWatcher(namespace, opts, onGetWatchError),
+		k.GetServiceWatcher(namespace, opts, onGetWatchError),
+		k.GetRouteWatcher(namespace, opts, onGetWatchError),
+	)
+	newWatch.SetFilters([]watch.EventType{watch.Added, watch.Modified, watch.Deleted})
+	newWatch.StartWatcher()
+
+	return newWatch
+}
+
+// Compile the metav1.ListOptions for resources.
+func getResourcesListOptions(nMetaMap map[string][]nodeMeta) map[metav1.ListOptions]nodeMeta {
+	listOptions := make(map[metav1.ListOptions]nodeMeta)
+
+	// For each node, create the metav1.ListOptions based off
+	// the app.kubernetes.io/name label.
+	for lKey, nMetas := range nMetaMap {
+		if lKey != "" {
+			options := metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app.kubernetes.io/name=%s", lKey),
+			}
+			for _, meta := range nMetas {
+				listOptions[options] = meta
+			}
+		}
+	}
+
+	return listOptions
+}
+
+// Gets node metadata.
+func getNodeMetadata(event watch.Event) nodeMeta {
+	var x interface{} = event.Object
+	var node nodeMeta
+	switch x.(type) {
+	case *deploymentconfigv1.DeploymentConfig:
+		dc := x.(*deploymentconfigv1.DeploymentConfig)
+		node = nodeMeta{
+			ID:          base64.StdEncoding.EncodeToString([]byte(dc.UID)),
+			Name:        dc.Name,
+			Kind:        "DeploymentConfig",
```

Comment:
Shouldn't `Kind` have `constant` values instead of `string`?

Reply:
Initially, thought we could get the kind by calling `.Kind` or `.GetObjectKind()` however, `.Kind` returns empty on all resources and `.GetObjectKind()` returns `&TypeMeta{Kind:,APIVersion:,}` 

I am unsure why or how else i can get the kind - do you know by any chance? I can see that it set in the yaml.